benches/engines/mod.rs

```diff
@@ -0,0 +1,232 @@
+        engine.snapshot(&ctx).is_ok();
+    })
+}
+
+/// Actually, it measures the performance of Snapshot::get(), skipping the Engine::snapshot();
+fn engine_get_bench<E: Engine>(
+    bencher: &mut Bencher,
+    engine: &E,
+    iterations: usize,
+    engine_keys_count: usize,
+    value_length: usize,
+) {
+    fill_engine_with(engine, engine_keys_count, value_length);
+    let ctx = Context::new();
+    let test_kvs = generate_random_kvs(iterations, DEFAULT_KEY_LENGTH, value_length);
+
+    bencher.iter_with_setup(
+        || engine.snapshot(&ctx).unwrap(),
+        |snap| {
+            for (key, _) in &test_kvs {
+                snap.get(&Key::from_raw(key)).is_ok();
+            }
+        },
+    );
+}
+
+/// Measure the performance of Engine::put()
+fn engine_put_bench<E: Engine>(
+    bencher: &mut Bencher,
+    engine: &E,
+    write_count: usize,
+    value_length: usize,
+) {
+    let test_kvs = generate_random_kvs(write_count, DEFAULT_KEY_LENGTH, value_length);
+    let ctx = Context::new();
+    bencher.iter(|| {
+        for (key, value) in &test_kvs {
+            let _ = engine.put(&ctx, Key::from_raw(&key), value.clone());
+        }
+    });
+}
+
+#[derive(Debug)]
+struct PutConfig {
+    engine_type: EngineType,
+    put_count: usize,
+    value_length: usize,
+}
+
+fn bench_engine_put(bencher: &mut Bencher, config: &PutConfig) {
+    match config.engine_type {
+        EngineType::BTreeEngine => {
+            let engine = BTreeEngine::default();
+            engine_put_bench(bencher, &engine, config.put_count, config.value_length)
+        }
+        EngineType::RocksDB => {
+            let engine = TestEngineBuilder::new().build().unwrap();
+            engine_put_bench(bencher, &engine, config.put_count, config.value_length)
+        }
+    }
+}
+
+#[derive(Debug)]
+struct SnapshotConfig {
+    engine_type: EngineType,
+    engine_keys_count: usize,
+    value_length: usize,
+}
+
+fn bench_engine_snapshot(bencher: &mut Bencher, config: &SnapshotConfig) {
+    match config.engine_type {
+        EngineType::BTreeEngine => {
+            let engine = BTreeEngine::default();
+            engine_snapshot_bench(
+                bencher,
+                &engine,
+                config.engine_keys_count,
+                config.value_length,
+            )
+        }
+        EngineType::RocksDB => {
+            let engine = TestEngineBuilder::new().build().unwrap();
+            engine_snapshot_bench(
+                bencher,
+                &engine,
+                config.engine_keys_count,
+                config.value_length,
+            )
+        }
+    }
+}
+
+#[derive(Debug)]
+struct GetConfig {
+    engine_type: EngineType,
+    iterations: usize,
+    value_length: usize,
+    engine_keys_count: usize,
+}
+
+fn bench_engine_get(bencher: &mut Bencher, config: &GetConfig) {
+    match config.engine_type {
+        EngineType::BTreeEngine => {
+            let engine = BTreeEngine::default();
```

Comment:
How about:

```rust
impl<E: Engine> Engine for Box<E> { .... }
```

Then:

```rust
#[derive(Debug, Clone, Copy)]
enum EngineBuilder {
    BTreeEngine,
    RocksDB,
}

impl EngineBuilder {
    pub fn build(&self) -> Box<Engine> {
        match self { ... }
    }
}

#[derive(Debug)]
struct GetConfig {
    engine_builder: EngineBuilder,
    iterations: usize,
    value_length: usize,
    engine_keys_count: usize,
}
```

So now, you don't need a bridge function that checks `engine_type` any more, because `engine_builder.build()` will build a `Box<Engine>`, which meets `E: Engine` trait bound.

Reply:
Got it!!!  I tried to drop the bridge function in many ways, except `box`!